bpo-37556 Extend help to include latest overrides

GadgetSteve opened the pull request:
Modify the help in cpython/PC/launcher.c to show users that "latest" can be overridden by shebang, PY_PYTHON[n] or py.ini files. Also show that script [args] is optional by enclosing in square brackets.

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
bpo-NNNN: Summary of the changes made
```

Where: bpo-NNNN refers to the issue number in the https://bugs.python.org.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `master`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `master`.

-->


<!-- issue-number: [bpo-37556](https://bugs.python.org/issue37556) -->
https://bugs.python.org/issue37556
<!-- /issue-number -->


Automerge-Triggered-By: @zooba

GadgetSteve committed:
bpo-37556 Extend help to include latest overrides

Modify the help in cpython/PC/launcher.c to show users that "latest" can be overridden by shebang, PY_PYTHON[n] or py.ini files. Also show that script [args] is optional.

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

zooba commented on PC/launcher.c:
```diff
@@ -1412,22 +1412,29 @@ show_help_text(wchar_t ** argv)
--2     : Launch the latest Python 2.x version\n\
--3     : Launch the latest Python 3.x version\n\
+-2     : Launch the latest* Python 2.x version\n\
+-3     : Launch the latest* Python 3.x version\n\
```
I think we can do without the footnote marker, as the comment below applies to practically all possibilities (and the text itself explains which cases are not covered).

GadgetSteve replied:
Addressed!

eryksun commented on PC/launcher.c:
```diff
@@ -1412,22 +1412,29 @@ show_help_text(wchar_t ** argv)
     fputws(L"\n-0  --list       : List the available pythons", stdout);
     fputws(L"\n-0p --list-paths : List with paths", stdout);
+    fputws(L"\n\n If no script is specified the specified interpreter is opened.", stdout);
+    fputws(L"\n*If an exact version is not given latest can be overriden by:", stdout);
```
This line is missing a comma and "overridden" is misspelled. I suggest the following: "If an exact version is not given, using the latest version can be overridden by any of the following:".

GadgetSteve replied:
Corrected and added that the overrides are in priority order.

GadgetSteve committed:
bpo-37556 Remove footnote marker

Removed the footnote `*` maker from the help lines as per comment from @zooba

eryksun commented on PC/launcher.c:
```diff
@@ -1412,22 +1412,29 @@ show_help_text(wchar_t ** argv)
+    fputws(L"\n*If an exact version is not given latest can be overriden by:", stdout);
+    fputws(L"\n An active virtual environment", stdout);
+    fputws(L"\n A shebang line in the script (if present)", stdout);
+    fputws(L"\n Matching PY_PYTHON[n] Enviroment variable(s)", stdout);
```
Do you think it's clear that `PY_PYTHON` sets the default and `PY_PYTHON2` and `PY_PYTHON3` are for the -2 and -3 options?

GadgetSteve replied:
Reworded to make clearer

eryksun commented on PC/launcher.c:
```diff
@@ -1412,22 +1412,29 @@ show_help_text(wchar_t ** argv)
+    fputws(L"\n A shebang line in the script (if present)", stdout);
+    fputws(L"\n Matching PY_PYTHON[n] Enviroment variable(s)", stdout);
+    fputws(L"\n From [defaults] in py.ini in your %LOCALAPPDATA%\\py.ini", stdout);
+    fputws(L"\n From [defaults] in py.ini beside py.exe (use where py to locate)", stdout);
```
Put \`where py\` in backticks to clarify the command here.

GadgetSteve replied:
Done

GadgetSteve committed:
bpo-37556 Address comments by Eryk Sun

Reworded and corrected spelling per comments by @eryksun also added that the overrides are in priority order.

eryksun commented on PC/launcher.c:
```diff
@@ -1412,22 +1412,31 @@ show_help_text(wchar_t ** argv)
+    fputws(L"\nany of the following, (in priority order):", stdout);
+    fputws(L"\n An active virtual environment", stdout);
+    fputws(L"\n A shebang line in the script (if present)", stdout);
+    fputws(L"\n Matching PY_PYTHON2 or 3 Enviroment variable", stdout);
```
Without reading the docs, I'd be inclined to think that `PY_PYTHON3` should override the latest Python 3 selection, but it only applies with the `-3` command-line option. The same for `PY_PYTHON2`. So maybe change this to 

> "A PY_PYTHON2 or PY_PYTHON3 environment variable in combination with the -2 or -3 argument, respectively."

GadgetSteve replied:
Modified to:
 `With -2 or -3 flag a matching PY_PYTHON2 or PY_PYTHON3 Enviroment variable`

_(Shorter than suggested but clearer than before)_.

GadgetSteve committed:
bpo-37556 Clarify PY_PYTHON2/3

To address commend from @eryksun reworded to:
 With -2 or -3 flag a matching PY_PYTHON2 or PY_PYTHON3 Enviroment variable

zooba commented on PC/launcher.c:
```diff
@@ -1412,22 +1412,31 @@ show_help_text(wchar_t ** argv)
     if (canDo64bit) {
         fputws(L"\
-     The above all default to 64 bit if a matching 64 bit python is present.\n\
+     The above all default to 64 bit if a matching 64 bit python is present*.\n\
```
```suggestion
     The above all default to 64 bit if a matching 64 bit python is present.\n\
```

Missed one

GadgetSteve replied:
Darn! Fixed.

GadgetSteve committed:
bpo-37556: Remove another asterisk

Missed one!
